neuralpp/symbolic/z3_expression.py

```diff
@@ -349,3 +365,118 @@ def subexpressions(self) -> List[Expression]:
     @property
     def number_of_arguments(self) -> int:
         return self.z3_object.num_args()
+
+
+z3_false = z3.BoolVal(False)
+
+
+def extract_key_value_from_assertions_helper(assertions: List[z3.ExprRef] | z3.AstVector, dict_: Dict[str, Any]):
+    def is_value(v):
+        return z3.is_int_value(v) or z3.is_rational_value(v) or z3.is_algebraic_value(v) or z3.is_fp_value(v)
+
+    def is_key(k):
+        return z3.is_const(k) and not is_value(k)
+
+    for assertion in assertions:
+        if z3.is_eq(assertion):
+            lhs, rhs = assertion.arg(0), assertion.arg(1)
+            if is_key(lhs) and is_value(rhs):
+                dict_[str(lhs)] = rhs
+            elif is_key(rhs) and is_value(lhs):
+                dict_[str(rhs)] = lhs
+        elif z3.is_and(assertion):
+            extract_key_value_from_assertions_helper(assertion.children(), dict_)
+
+
+def extract_key_value_from_assertions(assertions: List[z3.ExprRef] | z3.AstVector) -> Dict[str, Any]:
+    """
+    Extract all key-value pairs if it's in the form of "k1 == v1 & k2 == v2 & ..". `And` can be nested:
+    we can extract all 4 pairs in And(k1 == v1, k2 == v2, And(k3 == v3, k4 == v4)).
+    """
+    result = {}
+    extract_key_value_from_assertions_helper(assertions, result)
+    return result
+
+
+class Z3SolverExpression(Context, Z3Expression, FunctionApplication):
+    @property
+    def dict(self) -> Dict[str, Any]:
+        return self._dict
+
+    def __eq__(self, other) -> bool:
+        return False  # why do we need to compare two Z3ConjunctiveClause?
+
+    def __init__(self, z3_solver: z3.Solver, value_dict: Dict[str, Any] | None = None):
+        """
+        Assume z3_solver is satisfiable, otherwise user should use Z3UnsatContext() instead.
+        Also assumes value_dict is not contradictory to z3_solver. Formally, the following statement is valid:
+            for all k,v in value_dict.item(), z3_solver.assertions implies k == v
+        """
+        # assert z3_solver.check() == sat
+        super().__init__(z3_false)
+        self._solver = z3_solver
+        if value_dict is not None:
+            self._dict = value_dict
+        else:  # figure out ourselves
+            self._dict = extract_key_value_from_assertions(z3_solver.assertions())
+
+    @cached_property
+    def assertions(self) -> z3.AstVector:
+        return self._solver.assertions()
+
+    @property
+    def function(self) -> Expression:
+        # conjunctive clause is an "and" of all arguments
+        return Z3Constant(z3.And(True, True).decl())
+
+    @cached_property
+    def arguments(self) -> List[Expression]:
+        return [z3_object_to_expression(assertion) for assertion in self.assertions]
+
+    @property
+    def subexpressions(self) -> List[Expression]:
+        return [self.function] + self.arguments
+
+    @property
+    def unsatisfiable(self) -> bool:  # self should always be satisfiable
+        return False
```

Comment:
We can be sure about this because `Expression` is immutable. Whenever we `make()` an unsatisfiable context, we return `FalseContext()`. So `Z3SolverExpression()` created by our symbolic library is always satisfiable.

Though user needs to follow `__init__()`'s assumption and never passes a `unsat` solver to it. (We can check for `z3_solver.check() == sat` in `__init__()` but that would be some redundant computation to library calls.)

Reply:
On a second thought, it may be better to have this `check() == sat`, since it's good for correctness, and I believe z3 should have some caching mechanism if I `check()` a same solver twice.